clearwater-config-manager.root/usr/share/clearwater/clearwater-config-manager/scripts/log_shared_config.py

```diff
@@ -0,0 +1,90 @@
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version, along with the "Special Exception" for use of
+# the program along with SSL, set forth below. This program is distributed
+# in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import os
+import sys
+import json
+import difflib
+import subprocess
+import syslog
+
+def main():
+    '''
+    Print a readable diff of changes between etcd's shared_config and a newer
+    local copy, and log to syslog.
+    '''
+
+    # URL of shared_config etcd key
+    url = sys.argv[1]
+
+    # Get the old shared_config stored on etcd
+    # If this fails, this script is the least of our worries and we ignore, because
+    # upload_shared_config, which called this script, will print an error instead
+    jsonstr = subprocess.check_output(["curl", "-s", "-X", "GET", url])
+
+    new_config_lines = open("/etc/clearwater/shared_config").read().splitlines()
+    # etcd returns JSON; the shared_config is in node.value
+    old_config_lines = json.loads(jsonstr)["node"]["value"].splitlines()
+
+    # We're looking to log meaningful configuration changes, so sort the lines to
+    # ignore changes in line ordering
+    new_config_lines.sort()
+    old_config_lines.sort()
+    difflines = list(difflib.ndiff(old_config_lines, new_config_lines))
+
+    # (i) Pull out diff lines prefixed by "+ " / "- "; (ii) separate with quotes and commas;
+    # (iii) ignore empty lines; (iv) concatenate; (v) delete trailing comma
+    additions = ''.join(("\"" + line[2:] + "\", ") for line in difflines \
```

Comment:
I think it's probably more readable if you:
- split this up a bit (separating out "choosing the right lines" from "formatting it properly")
- use single quotes for Python strings containing double quotes, instead of escaping them
- join the entries with ", " instead of putting one after every entry and stripping it

e.g.

```
additions = [line[2:] for line in difflines if line.startswith("+ ") and len(line) > 2]
additions_str = ", ".join(['"{}"`.format(line) for line in additions]
```